src/lib/mgba/websocket-client.ts

```diff
@@ -0,0 +1,558 @@
+            !message.startsWith('{')) {
+          console.log('mGBA:', message)
+          return // Don't resolve/reject, wait for actual response
+        }
+
+        try {
+          const response = JSON.parse(message) as MgbaEvalResponse
+          this.ws?.off('message', messageHandler)
+          resolve(response)
+        } catch (error) {
+          this.ws?.off('message', messageHandler)
+          reject(new Error(`Failed to parse eval response: ${String(error)}`))
+        }
+      }
+
+      this.ws.on('message', messageHandler)
+
+      // Send the code to evaluate
+      this.ws.send(code)
+
+      // Timeout after 5 seconds
+      setTimeout(() => {
+        this.ws?.off('message', messageHandler)
+        reject(new Error('Eval request timed out'))
+      }, 5000)
+    })
+  }
+
+  /**
+   * Read a byte from memory
+   */
+  async readByte (address: number): Promise<number> {
+    const response = await this.eval(`emu:read8(${address})`)
+    if (response.error) {
+      throw new Error(`Failed to read byte at 0x${address.toString(16)}: ${response.error}`)
+    }
+    return response.result as number
+  }
+
+  /**
+   * Read a 16-bit value from memory (little-endian)
+   */
+  async readWord (address: number): Promise<number> {
+    const response = await this.eval(`emu:read16(${address})`)
+    if (response.error) {
+      throw new Error(`Failed to read word at 0x${address.toString(16)}: ${response.error}`)
+    }
+    return response.result as number
+  }
+
+  /**
+   * Read a 32-bit value from memory (little-endian)
+   */
+  async readDWord (address: number): Promise<number> {
+    const response = await this.eval(`emu:read32(${address})`)
+    if (response.error) {
+      throw new Error(`Failed to read dword at 0x${address.toString(16)}: ${response.error}`)
+    }
+    return response.result as number
+  }
+
+  /**
+   * Read multiple bytes from memory (OPTIMIZED VERSION)
+   * Uses the fastest available method: readRange API when possible, falls back to bulk Lua
+   */
+  async readBytes (address: number, length: number): Promise<Uint8Array> {
+    // Use native readRange API for maximum performance
+    return this.readBytesNative(address, length)
+  }
+
+  /**
+   * Read multiple bytes using mGBA's native readRange API (FASTEST)
+   * This is the fastest possible method using mGBA's built-in memory reading
+   */
+  async readBytesNative (address: number, length: number): Promise<Uint8Array> {
+    // For very small reads, use bulk Lua which is more reliable
+    if (length <= 10) {
+      return this.readBytesBulk(address, length)
+    }
+
+    try {
+      // Try using readRange API - it returns binary data as a string
+      const luaCode = `(function()
+        local data = emu:readRange(${address}, ${length})
+        local bytes = {}
+        for i = 1, #data do
+          bytes[i] = string.byte(data, i)
+        end
+        return bytes
+      end)()`
```

Comment:
lets change the http-server code, so that it can support non self executing function inputs